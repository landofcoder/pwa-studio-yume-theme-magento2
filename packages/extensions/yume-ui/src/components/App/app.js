import React, { useCallback } from 'react';
import { useIntl } from 'react-intl';
import { array, func, shape, string } from 'prop-types';

import { useToasts } from '@magento/peregrine';
import { useApp } from '@magento/peregrine/lib/talons/App/useApp';

import globalCSS from '@landofcoder/yume-ui/src/index.css';
import { HeadProvider, Title } from '@magento/venia-ui/lib/components/Head';
import Main from '@landofcoder/yume-ui/src/components/Main';
import Mask from '@magento/venia-ui/lib/components/Mask';
import Navigation from '@magento/venia-ui/lib/components/Navigation';
// import Routes from '@magento/venia-ui/lib/components/Routes';
import Routes from '@landofcoder/yume-ui/src/components/Routes';
import ToastContainer from '@magento/venia-ui/lib/components/ToastContainer';
import Icon from '@magento/venia-ui/lib/components/Icon';

import {
    AlertCircle as AlertCircleIcon,
    CloudOff as CloudOffIcon,
    Wifi as WifiIcon
} from 'react-feather';

const OnlineIcon = <Icon src={WifiIcon} attrs={{ width: 18 }} />;
const OfflineIcon = <Icon src={CloudOffIcon} attrs={{ width: 18 }} />;
const ErrorIcon = <Icon src={AlertCircleIcon} attrs={{ width: 18 }} />;

const App = props => {
    const { markErrorHandled, renderError, unhandledErrors } = props;
    const { formatMessage } = useIntl();

    const [, { addToast }] = useToasts();

    const ERROR_MESSAGE = formatMessage({
        id: 'app.errorUnexpected',
        defaultMessage: 'Sorry! An unexpected error occurred.'
    });

    const handleIsOffline = useCallback(() => {
        addToast({
            type: 'error',
            icon: OfflineIcon,
            message: formatMessage({
                id: 'app.errorOffline',
                defaultMessage:
                    'You are offline. Some features may be unavailable.'
            }),
            timeout: 3000
        });
    }, [addToast, formatMessage]);

    const handleIsOnline = useCallback(() => {
        addToast({
            type: 'info',
            icon: OnlineIcon,
            message: formatMessage({
                id: 'app.infoOnline',
                defaultMessage: 'You are online.'
            }),
            timeout: 3000
        });
    }, [addToast, formatMessage]);

    const handleError = useCallback(
        (error, id, loc, handleDismissError) => {
            const errorToastProps = {
                icon: ErrorIcon,
                message: `${ERROR_MESSAGE}\nDebug: ${id} ${loc}`,
                onDismiss: remove => {
                    handleDismissError();
                    remove();
                },
                timeout: 15000,
                type: 'error'
            };

            addToast(errorToastProps);
        },
        [ERROR_MESSAGE, addToast]
    );

    const talonProps = useApp({
        handleError,
        handleIsOffline,
        handleIsOnline,
        markErrorHandled,
        renderError,
        unhandledErrors
    });

    const { hasOverlay, handleCloseDrawer } = talonProps;

    if (renderError) {
        return (
            <HeadProvider>
                <Title>
                    {formatMessage(
                        { id: 'app.titleHome', defaultMessage: 'Home Page' },
                        { name: STORE_NAME }
                    )}
                </Title>
                <Main isMasked={true} />
                <Mask isActive={true} />
                <ToastContainer />
            </HeadProvider>
        );
    }

    return (
        <HeadProvider>
            <Title>
                {formatMessage(
                    { id: 'app.titleHome', defaultMessage: 'Home Page' },
                    { name: STORE_NAME }
                )}
            </Title>
            <Main isMasked={hasOverlay}>
                <Routes />
            </Main>
            <Mask isActive={hasOverlay} dismiss={handleCloseDrawer} />
            <Navigation />
            <ToastContainer />
        </HeadProvider>
    );
};

App.propTypes = {
    markErrorHandled: func.isRequired,
    renderError: shape({
        stack: string
    }),
    unhandledErrors: array
};

App.globalCSS = globalCSS;

export default App;
