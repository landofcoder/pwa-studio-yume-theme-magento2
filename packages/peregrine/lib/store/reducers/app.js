import { handleActions } from 'redux-actions';

import actions from '../actions/app';

export const name = 'app';

const initialState = {
    drawer: null,
    hasBeenOffline: !navigator.onLine,
    isOnline: navigator.onLine,
    isPageLoading: false,
    overlay: false,
    pending: {},
    searchOpen: false
};

const reducerMap = {
    [actions.toggleDrawer]: (state, { payload }) => {
        return {
            ...state,
            drawer: payload,
            overlay: !!payload
        };
    },
    [actions.toggleSearch]: state => {
        return {
            ...state,
            searchOpen: !state.searchOpen
        };
    },
    [actions.setOnline]: state => {
        return {
            ...state,
            isOnline: true
        };
    },
    [actions.setOffline]: state => {
        return {
            ...state,
            isOnline: false,
            hasBeenOffline: true
        };
    },
    [actions.setPageLoading]: (state, { payload }) => {
        return {
            ...state,
            isPageLoading: !!payload
        };
    }
};

export default handleActions(reducerMap, initialState);
