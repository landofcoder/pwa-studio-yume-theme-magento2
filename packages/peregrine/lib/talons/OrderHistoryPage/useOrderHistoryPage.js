import { useEffect, useMemo } from 'react';
import { useHistory } from 'react-router-dom';
import { useQuery } from '@apollo/client';

import { useAppContext } from '../../context/app';
import { useUserContext } from '../../context/user';
import { useTypePolicies } from '../../hooks/useTypePolicies';

export const useOrderHistoryPage = props => {
    const { queries, types } = props;
    const { getCustomerOrdersQuery } = queries;

    useTypePolicies(types);

    const [
        ,
        {
            actions: { setPageLoading }
        }
    ] = useAppContext();
    const history = useHistory();
    const [{ isSignedIn }] = useUserContext();

    const { data, loading } = useQuery(getCustomerOrdersQuery, {
        fetchPolicy: 'cache-and-network',
        skip: !isSignedIn
    });

    const isLoadingWithoutData = !data && loading;
    const isBackgroundLoading = !!data && loading;
    const orders = useMemo(() => {
        if (data) {
            return data.customer.orders.items;
        }

        return [];
    }, [data]);

    // If the user is no longer signed in, redirect to the home page.
    useEffect(() => {
        if (!isSignedIn) {
            history.push('/');
        }
    }, [history, isSignedIn]);

    // Update the page indicator if the GraphQL query is in flight.
    useEffect(() => {
        setPageLoading(isBackgroundLoading);
    }, [isBackgroundLoading, setPageLoading]);

    return {
        isLoadingWithoutData,
        orders
    };
};
