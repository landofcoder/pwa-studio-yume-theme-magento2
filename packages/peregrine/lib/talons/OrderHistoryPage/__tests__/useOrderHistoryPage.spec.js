import React, { useEffect } from 'react';
import { act } from 'react-test-renderer';
import { useQuery } from '@apollo/client';

import createTestInstance from '../../../util/createTestInstance';
import { useOrderHistoryPage } from '../useOrderHistoryPage';
import { useAppContext } from '../../../context/app';

jest.mock('react-router-dom', () => {
    return {
        useHistory: jest.fn(() => ({ push: jest.fn() }))
    };
});

jest.mock('@apollo/client', () => ({
    useQuery: jest.fn().mockReturnValue({
        data: {
            customer: {
                orders: {
                    items: ['order1', 'order2']
                }
            }
        },
        loading: false
    })
}));

jest.mock('@magento/peregrine/lib/context/app', () => {
    const state = {};
    const api = { actions: { setPageLoading: jest.fn() } };
    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

jest.mock('@magento/peregrine/lib/context/user', () => {
    const state = {
        isSignedIn: false
    };
    const api = {};
    const useUserContext = jest.fn(() => [state, api]);

    return { useUserContext };
});

jest.mock('../../../hooks/useTypePolicies', () => ({
    useTypePolicies: jest.fn()
}));

const log = jest.fn();
const Component = props => {
    const talonProps = useOrderHistoryPage({ ...props });

    useEffect(() => {
        log(talonProps);
    }, [talonProps]);

    return null;
};

const props = { queries: {} };

describe('it returns the proper shape', () => {
    test('with data', () => {
        createTestInstance(<Component {...props} />);

        const talonProps = log.mock.calls[0][0];
        expect(talonProps).toMatchSnapshot();
    });

    test('while loading without data', () => {
        useQuery.mockReturnValue({
            loading: true
        });

        createTestInstance(<Component {...props} />);

        const talonProps = log.mock.calls[0][0];
        expect(talonProps).toMatchSnapshot();
    });
});

test('syncs background loading state', () => {
    const data = {
        customer: {
            orders: {
                items: ['order1', 'order2']
            }
        }
    };
    useQuery.mockReturnValue({ data, loading: false }).mockReturnValueOnce({
        data,
        loading: true
    });

    const [, { actions }] = useAppContext();
    const { setPageLoading } = actions;

    const root = createTestInstance(<Component {...props} />);
    act(() => {
        root.update(<Component {...props} />);
    });

    expect(setPageLoading).toHaveBeenNthCalledWith(1, true);
    expect(setPageLoading).toHaveBeenNthCalledWith(2, false);
});
