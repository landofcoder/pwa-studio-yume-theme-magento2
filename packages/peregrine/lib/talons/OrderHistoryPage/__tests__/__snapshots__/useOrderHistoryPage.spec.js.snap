// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it returns the proper shape while loading without data 1`] = `
Object {
  "isLoadingWithoutData": true,
  "orders": Array [],
}
`;

exports[`it returns the proper shape with data 1`] = `
Object {
  "isLoadingWithoutData": false,
  "orders": Array [
    "order1",
    "order2",
  ],
}
`;
